### ISSUE TITLE
Using both getDerivedStateFromError and getDerivedStateFromProps can be a foot gun

### ISSUE THREAD

**Issue by blixt** [facebook/react#14049]
**Do you want to request a *feature* or report a *bug*?**

Feature

**What is the current behavior?**

If `componentDidCatch` and/or `getDerivedStateFromError` put the component in a state that avoids the cause of the error, and `getDerivedStateFromProps` reverts that state change, the error boundary will (obviously) fail to avert disaster.

https://codesandbox.io/s/pj0lwxk15j

It sounds very obvious when simplified like this, but when my team updated to React 16.5.x (and 16.6.0), suddenly this started happening for us. So something changed internally, but it's hard to pinpoint what. Unfortunately I haven't been able to create a small repro for that specific case that works in 16.4.2 but not 16.5.0 – so I'm making this issue about avoiding the cause in the first place.

**What is the expected behavior?**

Naively, my thinking is that because errors are more exceptional, let the state from error handlers take precedence. Probably there's a reason why this can't happen so a warning of some kind when this can happen would be nice. Two ways to do this comes to mind:

1. Cross-check that keys in state from `getDerivedStateFromProps` or `setState` in `componentDidCatch` don't collide with `getDerivedStateFromError` (sounds a bit far fetched?)
2. Be more explicit in documentation for error boundaries about how `getDerivedStateFromProps` will run last so that a developer can consider this case

**Which versions of React, and which browser / OS are affected by this issue? Did this work in previous versions of React?**

16.5.0 and up (16.4.2 behaves a bit differently – see above), all platforms/browsers.

**Comment by gaearon** [comment]
Hmm. I don't quite see why this is unexpected.

It seems to me like resetting the error in `getDerivedStateFromProps` is generally a bad idea, no? Unless you compare the props and *decide* to reset the error (e.g. maybe route changed). In which case I'd actually expect it to work.

Am I missing some practical scenario?

**Comment by blixt** [comment]
I think I agree, but since this caused a production crasher from updating a minor version of React I wanted to report it to see if there's anything that can be done to avoid it, but might also just be legacy code breaking due to intentional changes in React.

Basically, what happened in our case is that `getDerivedStateFromProps` was relied upon to not trigger twice in a row if an error occurred while rendering bu `children` (the only value that *can* change) did not change. Since 16.4 it changed (I just found #12898 so I see this has been discussed at length) and now `getDerivedStateFromProps` can be called twice due to a single update if an error occurs in the first render. Since `error` is reset in the second `getDerivedStateFromProps`, there's never a `render()` call that will render with it set to something, now causing two crashes in a row, which in turn causes the error boundary to be skipped, causing React to unmount entirely.

**Comment by gaearon** [comment]
I guess what I’m trying to say is that any code that returns something unconditionally from `getDerivedStateFromProps` is a symptom of a bug. It means your code relies on *when* updates happen rather than *what* changes.

**Comment by blixt** [comment]
Fair enough, it should definitely rely on what changes. Given the following scenario, what's the correct way to set `this.state.error` to `null` when `this.props.children` changes (eg a property change)?

```javascript
static getDerivedStateFromError(error) {
    return { error };
}

render() {
    if (this.state.error) {
        // Child component is in error state, show red box.
        return <RedBox />;
    }
    return this.props.children;
}
```

Here's a sandbox: https://codesandbox.io/s/035nw47kkp

(This indicates a bug in the child component, but it's a component out of our control and our UI will need to handle it and for good user experience recover once the component has been replaced or updated to work again.)

The options I can think of are:

* Find every place that the error boundary is used and imperatively tell it to reset its error before rendering (since if there's an error the `setState` will only happen once at this level, even as the component renders twice to recover from the error).
* Rely on the user to press a "Retry" button in the error boundary.
* Compare the `this.props.children` value that the component had upon error with the current value upon render and only show the error if it's identity equal to the erroring `this.props.children`.
* Some other way to attempt a re-render (eg by resetting `error`) when something further up the stack changes to probe `this.props.children` for potential errors and either recover from or retain the error state.

**Comment by EECOLOR** [comment]
I am not sure if this adds to the discussion

We have a similar situation situation where a third party component throws an error during render. Throwing errors during the render phase is bad, but it is not something we can fix in the short term.

The error comes up in rare occasions as a result of user input.

So our idea is: when the error is thrown we catch it, reset the user input and rerender the component.

Simplified it would look like this:

```jsx
class ErrorBoundary extends React.Component {
  static getDerivedStateFromError(e) {
    return { hasError: true }
  }
  state = { hasError: false }
  render() {
    return !this.state.hasError && this.props.children
  }
  componentDidCatch(e) {
    this.props.onError(e)
    this.setState({ hasError: false }) // the documentation states this will be deprecated
  }
}
```

It would however be better if we could tell this component that a recovery took place and it can safely resume its render duties.

**Comment by tonypee** [comment]
I would like to have the componentDidCatch cause an error toast message to occur, but have the application continue rendering. Is it possible to resume the render? 
I understand that i will need to resolve the state issues myself
